packages/shared/src/telemetry/collector.ts

```diff
@@ -0,0 +1,300 @@
+   */
+  cv: string;
+  /**
+   * SDK
+   */
+  sdk?: string;
+  /**
+   * SDK Version
+   */
+  sdkv?: string;
+  payload: Record<string, string | number | boolean>;
+};
+
+const DEFAULT_CONFIG: Partial<Required<TelemetryCollectorConfig>> = {
+  samplingRate: 1,
+  maxBufferSize: 5,
+  // Production endpoint: https://clerk-telemetry.com
+  // Staging endpoint: https://staging.clerk-telemetry.com
+  // Local: http://localhost:8787
+  endpoint: 'https://clerk-telemetry.com',
+};
+
+export class TelemetryCollector {
+  #config: Required<TelemetryCollectorConfig>;
+  #metadata: TelemetryMetadata = {} as TelemetryMetadata;
+  #buffer: TelemetryEvent[] = [];
+  #pendingFlush: any;
+
+  constructor(options: TelemetryCollectorOptions) {
+    this.#config = {
+      maxBufferSize: options.maxBufferSize ?? DEFAULT_CONFIG.maxBufferSize,
+      samplingRate: options.samplingRate ?? DEFAULT_CONFIG.samplingRate,
+      disabled: options.disabled ?? false,
+      debug: options.debug ?? false,
+      endpoint: DEFAULT_CONFIG.endpoint,
+    } as Required<TelemetryCollectorConfig>;
+
+    if (!options.clerkVersion && typeof window === 'undefined') {
+      // N/A in a server environment
+      this.#metadata.clerkVersion = '';
+    } else {
+      this.#metadata.clerkVersion = options.clerkVersion ?? '';
+    }
+
+    // We will try to grab the SDK data lazily when an event is triggered, so it should always be defined once the event is sent.
+    this.#metadata.sdk = options.sdk!;
+    this.#metadata.sdkVersion = options.sdkVersion!;
+
+    this.#metadata.publishableKey = options.publishableKey ?? '';
+
+    const parsedKey = parsePublishableKey(options.publishableKey);
+    if (parsedKey) {
+      this.#metadata.instanceType = parsedKey.instanceType;
+    }
+
+    if (options.secretKey) {
+      // Only send the first 16 characters of the secret key to to avoid sending the full key. We can still query against the partial key.
+      this.#metadata.secretKey = options.secretKey.substring(0, 16);
+    }
+  }
+
+  get isEnabled(): boolean {
+    if (this.#metadata.instanceType !== 'development') {
+      return false;
+    }
+
+    // In browser or client environments, we most likely pass the disabled option to the collector, but in environments
+    // where environment variables are available we also check for `CLERK_TELEMETRY_DISABLED`.
+    if (this.#config.disabled || (typeof process !== 'undefined' && isTruthy(process.env.CLERK_TELEMETRY_DISABLED))) {
+      return false;
+    }
+
+    return true;
+  }
+
+  get isDebug(): boolean {
+    return this.#config.debug || (typeof process !== 'undefined' && isTruthy(process.env.CLERK_TELEMETRY_DEBUG));
+  }
+
+  record(event: Pick<TelemetryEvent, 'event' | 'payload'>): void;
+  record(event: TelemetryEvent['event'], payload: TelemetryEvent['payload']): void;
+  record(
+    event: TelemetryEvent['event'] | Pick<TelemetryEvent, 'event' | 'payload'>,
+    payload?: TelemetryEvent['payload'],
+  ): void {
```

Comment:
❓ Why do we need multiple signatures here?

Another pattern I like that helps to eliminate the `|` unions for the final signature is simply typing the args for the final signature as `...args: unknown[]`. You only need to make the type wide enough to accept every possible type of the args, and `unknown[]` is always wide enough :)

Reply:
This let's us both pass a complete event, or use the more a more succinct form:
```ts
// full event
telemetry.record({ event: 'EVENT', payload: { ... } })

// short form
telemetry.record('EVENT', { ... })
```

In practice, the full event form is helpful when using the factory methods:
```ts
telemetry.record(eventComponentMounted(...))
```

> Another pattern I like that helps to eliminate the | unions for the final signature is simply typing the args for the final signature as ...args: unknown[]. You only need to make the type wide enough to accept every possible type of the args, and unknown[] is always wide enough :)

Is there a reason we might want to eliminate the union in the final signature? 🤔 I don't necessarily see a benefit to making that type wider in this scenario.

Reply:
Is there a reason we want to support the brief form? 
I prefer to limit the users of this API to the events created in the shared package for improved consistency, easier search replace, navigation, intellisense etc.

Reply:
yeah that's reasonable! I'll simplify this. Thanks for the feedback 🙂